### PR TITLE
feature: pictureInpicture

### PR DESCRIPTION
**Pull request by lisr2016**

### CODE REVIEW

File: src/component/types.ts [hunk 1]
```diff
@@ -1,11 +1,33 @@
-import { IsBoolean, IsEnum, IsNotEmpty, IsNumber, IsOptional, IsString, ValidateNested } from 'class-validator';
+import {IsArray, IsBoolean, IsEnum, IsNotEmpty, IsNumber, IsOptional, IsString, ValidateNested} from 'class-validator';
```
Comment (MengLi619) [inline src/component/types.ts:1]: Keep code format consistancy.

File: src/component/types.ts [hunk 2]
```diff
@@ -1,11 +1,33 @@
...
+export enum PipType {
+  bottomRight = 'bottom-right',
+  sideBySide = 'side-by-side',
+  threePictures = 'three-pictures',
+  fourCorners = 'four-corners',
+}
+
+export interface ObsPip {
+    backgroundPic?: string | null;
+    id:string;
```
Comment (MengLi619) [inline src/component/types.ts:18]: Code format

File: src/component/types.ts [hunk 3]
```diff
@@ -353,6 +414,8 @@ export interface Source {
   index: number;
   name: string;
   type: SourceType;
+  pipType?: PipType;
+  pipIndex?: number;
```
Comment (MengLi619) [inline src/component/types.ts:418]: Where do we use these two properties?

**Review by MengLi619 — CHANGES_REQUESTED** [review]

File: src/component/obs.ts [hunk 4]
```diff
@@ -178,7 +231,7 @@ export async function updateSource(source: Source, preview: Preview): Promise<vo
 }
 
 export function deleteSource(sourceId: string) {
-  obs.removeScene(sourceId);
```
[the remark this answers is not in the document]
Comment (lisr2016) [reply src/component/obs.ts:181]: mistaken

File: src/controller/pictureinpicturecontroller.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      await updatePictureInPicture(pip).catch(e => `Failed to update pictureInPicture ${pip.id}: ${e}`);
+    return pip;
+  }
+
+  @Delete('/v1/pictureInPictures/:pictureInPictureId')
+  @OnUndefined(StatusCodes.NO_CONTENT)
+  @Authorized(AUTHORIZED_ROLES)
+  public async delete(@Param('pictureInPictureId') pictureInPictureId: string): Promise<void> {
+    await deleteSwitcherPictureInPicture(pictureInPictureId);
+    deleteSource(pictureInPictureId);
```
[the remark this answers is not in the document]
Comment (lisr2016) [reply src/controller/pictureinpicturecontroller.ts:38]: corrected

File: src/component/types.ts [hunk 6]
```diff
@@ -172,6 +194,45 @@ export class UpdateAudioRequest {
...
+export class PipItem {
+  name:string;
+  index: number;
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+  static url: string;
+  url: string;
+}
```
[the remark this answers is not in the document]
Comment (lisr2016) [reply src/component/types.ts:220]: x,y,width,height 是为后面的自定义画中画准备的

**Pull request closed without merging**